ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c

```diff
@@ -387,44 +633,323 @@ dbmdb_import_add_created_attrs(Slapi_Entry *e)
+        "DNRC_NODN", "DNRC_NOPARENT_DN", "DNRC_NOPARENT_ID", "DNRC_NORDN",
+        "DNRC_RUV", "DNRC_SUFFIX", "DNRC_TOMBSTONE", "DNRC_VERSION", "DNRC_WAIT"
+    };
+    return t[dnrc];
+}
+
+dnrc_t
+get_entry_type(WorkerQueueData_t *wqelmt, Slapi_DN *sdn)
+{
+    backend *be = wqelmt->winfo.job->inst->inst_be;
+    int len = SLAPI_ATTR_UNIQUEID_LENGTH;
+    const char *ndn = slapi_sdn_get_ndn(sdn);
+
+    if (slapi_be_issuffix(be, sdn)) {
+        return DNRC_SUFFIX;
+    }
+    if (PL_strncasecmp(ndn, SLAPI_ATTR_UNIQUEID, len) || ndn[len] != '=') {
+            return DNRC_OK;
+    }
+    /* Maybe a tombstone or a ruv */
+    if (wqelmt->datalen) {
+        /* Check objectclass */
+        char *pt0, *pt1, *pt2;
+        int len2 = (sizeof SLAPI_ATTR_OBJECTCLASS) -1;
+        for (pt0 = pt1 = wqelmt->data; (pt1 = strcasestr(pt1, ": " SLAPI_ATTR_VALUE_TOMBSTONE "\n")); pt1++) {
+            /* Get start of line */
+            for (pt2 = pt1; pt2 >= pt0 && *pt2 != '\n'; pt2--);
+            pt2++;
+            if (PL_strncasecmp(pt2, SLAPI_ATTR_OBJECTCLASS, len2)) {
+                continue;
+            }
+            if (pt2[len2] != ';' && pt2[len2] != ':') {
+                continue;
+            }
+            break;  /* Found it */
+        }
+        if (!pt1) {
+            return DNRC_OK;
+        }
+    } else {
+        struct backentry *ep = wqelmt->data;
+        if (!slapi_entry_flag_is_set(ep->ep_entry, SLAPI_ENTRY_FLAG_TOMBSTONE)) {
+            return DNRC_OK;
         }
     }
-    return 1;
+    if (0 != PL_strncasecmp(&ndn[len+1], RUV_STORAGE_ENTRY_UNIQUEID, sizeof(RUV_STORAGE_ENTRY_UNIQUEID) - 1)) {
+        return DNRC_TOMBSTONE;
+    } else {
+        return DNRC_RUV;
+    }
 }
 
-/* Set state when a job thread finish */
-static inline void __attribute__((always_inline))
-info_set_state(ImportWorkerInfo *info)
+/*
+ * Compute
+ * key -> entryinfo for both entry and parent entry.
+ * param->ekey and param->pkey should be set by call in EIP_RDN case otherwise they are computed from the sdn
+ * ( and then point to sdn data so their is no need to freed them )
+ */
+static dnrc_t
+dbmdb_import_entry_info_by_param(EntryInfoParam_t *param, WorkerQueueData_t *wqelmt)
 {
-    if (info->state & ABORTED) {
-        info->state = FINISHED | ABORTED;
+    MDB_val data = {0};
+    Slapi_RDN srdn = {0};
+    const char *rdn = NULL;
+    const char *nrdn = NULL;
+    dnrc_t dnrc = DNRC_OK;
+    size_t len = 0;
+    static const ID pinfozero[6] = {0};
+    const ID *pinfo = pinfozero;
+    int pidfromkey = 0;
+    int rc = 0;
+
+    wqelmt->parent_info = NULL;
+    wqelmt->entry_info = NULL;
+    if (param->flags & EIP_RDN) {
+        memcpy(&pidfromkey, param->pkey.mv_data, sizeof (ID));
+    }
+
+    dnrc = get_entry_type(wqelmt, &param->sdn);
+    if (dnrc == DNRC_SUFFIX) {
+        if ( param->eid != 1) {
+            dnrc = DNRC_BAD_SUFFIX_ID;
+        } else {
+            rdn = slapi_sdn_get_dn(&param->sdn);
```

Comment:
This call returns a 'dn' not a 'rdn'. should not it call slapi_sdn_get_rdn ?

Reply:
Short answer is no. (suffix "rdn" is the dn) :
"rdn" notion in "entryrdn" index is a bit confusing:
  although it is the rdn for standard and RUV entry 
   it is the dn for the suffix and the 2 first rdns for tombstone entries (RUV excepted)